src/elfpy/markets.py

```diff
@@ -377,23 +470,148 @@ def _close_long(self, trade_details):
+                trade_details.mint_time,
+                trade_details.wallet_address,
+                - trade_details.trade_amount
+                ]
+        )
+        agent_deltas = AgentWallet( # write out explicit signs, so it's clear what's happening
+            base_in_wallet              = + d_base_reserves,
+            lp_in_wallet                = - lp_in,
+        )
+        return market_deltas, agent_deltas
+
+    def calc_fees_owed(self, return_agent=None):
+        """
+        Returns the fees owed to the agent
+        """
+        start_time = time.time()
+        cum_liq, prev_time, prev_share_fees, prev_token_fees = 0,0,0,0
+        cum_liq_by_agent, contribution, token_owed,share_owed = {},{},{},{}
+        for [current_time, acting_agent, new_liq] in self.liquidity_pool_history:
+            # calculate what happened since the last update, we use marginal values for everything
+            share_fees_till_now = sum([v for k,v in self.share_fee_history.items() if k <= current_time])
+            token_fees_till_now = sum([v for k,v in self.token_fee_history.items() if k <= current_time])
+            delta_share_fees = share_fees_till_now - prev_share_fees
+            delta_token_fees = token_fees_till_now - prev_token_fees
+            delta_time = current_time - prev_time
+            # initialize agent if this is the first time we see them
+            if acting_agent not in cum_liq_by_agent:
+                contribution[acting_agent], token_owed[acting_agent], share_owed[acting_agent],cum_liq_by_agent[acting_agent] = 0,0,0,0
+            if current_time != 0:  # only do a marginal update after first timestep where deltas are zero
+                for update_agent in cum_liq_by_agent:  # for each agent
+                    # update their marginal share of cumulative liquidity, give them credit for it (contribution)
+                    contribution[update_agent] += cum_liq_by_agent[update_agent] / cum_liq
+                    # update their owed fees
+                    share_owed[update_agent] = (share_owed[update_agent]*prev_time + contribution[update_agent]*delta_share_fees*delta_time)/current_time
+                    token_owed[update_agent] = (token_owed[update_agent]*prev_time + contribution[update_agent]*delta_token_fees*delta_time)/current_time
+            # update values used for next iteration
+            cum_liq += new_liq
+            cum_liq_by_agent[acting_agent] += new_liq
+            prev_time = current_time
+            prev_share_fees, prev_token_fees = share_fees_till_now, token_fees_till_now
+        if self.verbose:
+            print(f"cum_liq_by_agent = {cum_liq_by_agent}")
+            print(f"      share_owed = {share_owed} calculated sum = {sum(share_owed.values())}"
+                + f"\n                                           vs. direct = {self.share_fees} it's a "
+                + f"match 😁" if sum(share_owed.values()) == self.share_fees else f"mismatch 😢")
```

Comment:
I do not want to include emojis in the terminal outputs.
1) it is a branding decision that we haven't discussed
2) it requires checking to ensure that it is supported across all terminal types for our end-users, which doesn't exist and we don't have time to make.

If you'd like to have the discussion, you should make a separate PR that adds emojis.

Reply:
haha, imagine that PR 🙃
they're nice short-hand (i like making everything short)
but agree it was a pain in the ass to make them show up properly